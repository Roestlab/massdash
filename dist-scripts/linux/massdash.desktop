[Desktop Entry]
Version=0.0.8
Type=Application
Name=massdash
Exec=/usr/local/bin/massdash
Icon=/usr/share/icons/massdash.ico
Terminal=true
