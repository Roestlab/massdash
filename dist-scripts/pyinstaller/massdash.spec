# -*- mode: python ; coding: utf-8 -*-

import site
import importlib.metadata
import os
import sys

from PyInstaller.building.build_main import Analysis, PYZ, EXE, COLLECT
from PyInstaller.utils.hooks import collect_all, collect_data_files, copy_metadata

from transformers.dependency_versions_check import pkgs_to_check_at_runtime

sys.setrecursionlimit(sys.getrecursionlimit() * 5)

##################### User definitions
exe_name = 'massdash_gui'
script_name = 'run.py'
if sys.platform[:6] == "darwin":
	icon = '../../massdash/assets/img/MassDash_Logo.icns'
else:
	icon = '../../massdash/assets/img/MassDash_Logo.ico'
block_cipher = None
location = os.getcwd()
project = "massdash"
remove_tests = True
bundle_name = "massdash"
#####################


requirements = {project, "streamlit_javascript", "upsetplot", "distributed"}
datas = [(f"{site.getsitepackages()[0]}/streamlit/runtime", "./streamlit/runtime")]
hidden_imports = set()
binaries = []
checked = set()
while requirements:
	requirement = requirements.pop()
	print(f"Info: Checking {requirement}")
	checked.add(requirement)
	if requirement in ["pywin32"]:
		continue
	try:
		module_version = importlib.metadata.version(requirement)
	except (
		importlib.metadata.PackageNotFoundError,
		ModuleNotFoundError,
		ImportError
	):
		continue
	try:
		datas_, binaries_, hidden_imports_ = collect_all(
			requirement,
			include_py_files=True
		)
	except ImportError:
		continue
	datas += datas_
	# binaries += binaries_
	hidden_imports_ = set(hidden_imports_)
	if "" in hidden_imports_:
		hidden_imports_.remove("")
	if None in hidden_imports_:
		hidden_imports_.remove(None)
	requirements |= hidden_imports_ - checked
	hidden_imports |= hidden_imports_

if remove_tests:
	hidden_imports = sorted(
		[h for h in hidden_imports if "test" not in h.split(".")]
	)
else:
	hidden_imports = sorted(hidden_imports)


hidden_imports = [h for h in hidden_imports if "__pycache__" not in h]
datas = [d for d in datas if ("__pycache__" not in d[0]) and (d[1] not in [".", "Resources", "scripts"])]

#if sys.platform[:5] == "win32":
#	base_path = os.path.dirname(sys.executable)
#	library_path = os.path.join(base_path, "Library", "bin")
#	dll_path = os.path.join(base_path, "DLLs")
#	libcrypto_dll_path = os.path.join(dll_path, "libcrypto-1_1-x64.dll")
#	libssl_dll_path = os.path.join(dll_path, "libssl-1_1-x64.dll")
#	libcrypto_lib_path = os.path.join(library_path, "libcrypto-1_1-x64.dll")
#	libssl_lib_path = os.path.join(library_path, "libssl-1_1-x64.dll")
#	if not os.path.exists(libcrypto_dll_path):
#		datas.append((libcrypto_lib_path, "."))
#	if not os.path.exists(libssl_dll_path):
#		datas.append((libssl_lib_path, "."))

for _pkg in ["python","accelerate"]:
	if _pkg in pkgs_to_check_at_runtime:
		pkgs_to_check_at_runtime.remove(_pkg)
for _pkg in pkgs_to_check_at_runtime:
	datas += copy_metadata(_pkg)


datas += collect_data_files("streamlit")
datas += copy_metadata("streamlit")
datas += collect_data_files("pyopenms")
datas += copy_metadata("pyopenms")
datas += collect_data_files("massdash")
datas += copy_metadata("massdash")

hidden_imports = ['pyopenms', 'massdash', 'torchaudio.lib.libtorchaudio']

a = Analysis(
	[script_name],
	pathex=[location],
	binaries=binaries,
	datas=datas,
	hiddenimports=hidden_imports,
	hookspath=['./hooks'],
	runtime_hooks=[],
	excludes=[],
	win_no_prefer_redirects=False,
	win_private_assemblies=False,
	cipher=block_cipher,
	noarchive=False
)
pyz = PYZ(
	a.pure,
	a.zipped_data,
	cipher=block_cipher
)

if sys.platform[:5] == "linux":
	exe = EXE(
		pyz,
		a.scripts,
		a.binaries,
		a.zipfiles,
		a.datas,
		name=bundle_name,
		debug=False,
		bootloader_ignore_signals=False,
		strip=False,
		upx=True,
		console=True,
		upx_exclude=[],
		icon=icon
	)
else:
	exe = EXE(
		pyz,
		a.scripts,
		# a.binaries,
		a.zipfiles,
		# a.datas,
		exclude_binaries=True,
		name=exe_name,
		debug=False,
		bootloader_ignore_signals=False,
		strip=False,
		upx=True,
		console=True,
		icon=icon
	)
	coll = COLLECT(
		exe,
		a.binaries,
		# a.zipfiles,
		a.datas,
		strip=False,
		upx=True,
		upx_exclude=[],
		name=exe_name
	)