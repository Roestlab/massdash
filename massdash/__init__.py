"""
massdash
~~~~~~~~
TODO: ADD general documentation about how to use the auto API here

Package Structure
-----------------
testing - a test package
"""

__all__ = [ 'structs', 'preprocess', 'loaders', 'ui', 'server', 'testing' ]
__version__ = "0.0.9"
