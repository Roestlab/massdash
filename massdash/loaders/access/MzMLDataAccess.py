"""
massdash/loaders/access/MzMLDataAccess
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

import click
from typing import List, Tuple, Literal
import itertools
from tqdm import tqdm
from joblib import Parallel, delayed
import traceback
import sys
import mmap

# data modules
import pyopenms as po
import pandas as pd
import numpy as np

# Structs
from ...structs.TargetedDIAConfig import TargetedDIAConfig
from ...structs.FeatureMap import FeatureMap
from ...structs.TransitionGroupFeature import TransitionGroupFeature
# Internal
from ...util import LOGGER, method_timer, code_block_timer

class MzMLDataAccess():
    """
    Class for data loading and extracting operations of an .mzML file
    
    Attributes:
        filename (str): The mzML file to load.
        readOptions (str): The readOptions to use, either 'ondisk' or 'cached'.
        exp (OnDiscMSExperiment): The on disk experiment.
        meta_data (MSExperiment): The meta data.
        has_im (bool): Whether the data has ion mobility.
        
    Methods:
        load_data: Loads data from an mzML file as an on disc experiment for memory efficiency and meta data access without loading full data.
        get_target_ms_level_indices: Extract spectrum indices for a specific mslevel(s).
        get_spectra_rt_list: Get a list of RT for all the spectra using meta_exp.
    """

    def __init__(self, filename: str, readOptions="ondisk", verbose=False):
        """
        Initialise mzMLLoader object

        Args:
          mzml_file: (str) mzML file to load
          readOptions: (str) readOptions to use, either 'ondisk' or 'cached'
          verbose (bool): Enables verbose mode.
        """
        
        self.filename = filename
        self.readOptions = readOptions
        self.exp = po.OnDiscMSExperiment()
        self.meta_data = po.MSExperiment()
        self.has_im = self.check_ion_mobility()
        
        LOGGER.name = self.__class__.__name__
        if verbose:
            LOGGER.setLevel("DEBUG")
        else:
            LOGGER.setLevel("INFO")

        
        self.load_data()

    def __str__(self):
        return f"MzMLDataAccess(filename={self.filename}, has_im={self.has_im})"
 
    def __repr__(self):
        return f"MzMLDataAccess(filename={self.filename}, has_im={self.has_im})"
    
    def check_ion_mobility(self, num_lines_to_check=10_000_000):
        """
        Check if the mzML file contains ion mobility data

        Args:
        mzml_file: (str) mzML file to load
        num_lines_to_check: (int) Number of lines to check for "Ion Mobility"

        Returns:
        Return a boolean indicating if the mzML file contains ion mobility data
        """
        has_ion_mobility = False
        mzml_file = self.filename
        with open(mzml_file, 'rb', 0) as file:
            s = mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) 
            for line in iter(s.readline, b""):
                if b'Ion Mobility' in line:
                    has_ion_mobility = True
                    break 
                if s.tell() > num_lines_to_check:
                    break
        return has_ion_mobility

    @method_timer
    def load_data(self):
        """
        Method to load data from an mzML file as an on disc experiment for memory efficiency and meta data access without loading full data
        """

        if self.readOptions=="ondisk":
            with code_block_timer('Creating OnDiscExperiment...', LOGGER.debug):
                exp = po.OnDiscMSExperiment()

            with code_block_timer(f'Opening {self.filename} file...', LOGGER.info):
                exp.openFile(self.filename)

            with code_block_timer('Extracting meta data...', LOGGER.debug):
                meta_data = exp.getMetaData()

            LOGGER.info(
                f"There are {meta_data.getNrSpectra()} spectra and {exp.getNrChromatograms()} chromatograms.")
            LOGGER.info(
                f"There are {sum([spec.getMSLevel()==1  for spec in meta_data.getSpectra()])} MS1 spectra and {sum([spec.getMSLevel()==2  for spec in meta_data.getSpectra()])} MS2 spectra.")
        elif self.readOptions=="cached":
            # Because data is cached, we need to make an assumption about whether the file is MS1 or MS2
            with code_block_timer(f'Loading Cached Data from {self.filename} file...', LOGGER.info):
                exp = po.SpectrumAccessOpenMSCached(self.filename)
            ## Still need access to the meta data, so we need to create an on-disk experiment
            with code_block_timer('Creating OnDiscExperiment...', LOGGER.debug):
                od_exp = po.OnDiscMSExperiment()
            with code_block_timer(f'Opening {self.filename} file to extract meta-data...', LOGGER.info):
                od_exp.openFile(self.filename)
            with code_block_timer(f'Extracting meta data...', LOGGER.debug):
                meta_data = od_exp.getMetaData()
            # Delete OnDiskExperiment since we don't need it anymore, we only need the meta data
            del od_exp
    
            LOGGER.info(
                f"There are {exp.getNrSpectra()} spectra and {exp.getNrChromatograms()} chromatograms.")
            LOGGER.info(
                f"There are {sum([spec.getMSLevel()==1  for spec in meta_data.getSpectra()])} MS1 spectra and {sum([spec.getMSLevel()==2  for spec in meta_data.getSpectra()])} MS2 spectra.")
        else:
            click.ClickException(f"ERROR: Unknown readOptions ({self.readOptions}) given! Has to be one of 'ondisk', 'cached'")

        self.exp = exp
        self.meta_data = meta_data
            
    def get_target_ms_level_indices(self, mslevel=[1,2]) -> np.array:
        """
        Extract spectrum indices for a specific mslevel(s).

        Args:
          self: (object) self object containing meta_data
          mslevel: (list) list of mslevel(s) to extract indices for

        Return:
          Return mslevel_indices a list of indices with request mslevel(s) to self
        """
        if self.readOptions=="ondisk":
            spectra = self.meta_data.getSpectra()
            return np.array([indice for indice, spec in enumerate(
                spectra) if spec.getMSLevel() in mslevel])
        elif self.readOptions=="cached":
            return np.array(range(0,self.exp.getNrSpectra()))
        else:
            click.ClickException(f"ERROR: Unknown readOptions ({self.readOptions}) given! Has to be one of 'ondisk', 'cached'")

    def get_spectra_rt_list(self) -> np.array:
        """
        Get a list of RT for all the spectra using meta_exp

        Args:
          self: (object) self object containing meta_data

        Return:
          Return a list of RT values for spectra
        """
        if self.readOptions=="ondisk":
            meta_rt_list = np.array([meta_spec.getRT() for meta_spec in self.meta_data.getSpectra()])
        else: # self.readOptions=="cached"
            meta_rt_list = np.array([meta_spec.getRT()for meta_spec in self.meta_data.getSpectra()])
        return meta_rt_list
    
    def load_spectrum(self, spec_indice: int) -> Tuple[np.array, np.array, np.array]:
        """
        Load a single spectrum for a given spectrum indice.

        Args:
          spec_indice: (int) an interger of the spectrum indice to extra a spectrum for

        Return:
          mz_array: mz array
          int_array: intensity array
          im_array: ion mobility array 
        """
        # Get data arrays
        if (self.readOptions=="ondisk"):
            with code_block_timer(f'Extracting mz, int, im data arrays...', LOGGER.debug):
                spec = self.exp.getSpectrum(spec_indice)
                mz_array = spec.get_peaks()[0]
                int_array = spec.get_peaks()[1]
                if self.has_im:
                    im_array = spec.getFloatDataArrays()
                else:
                    im_array = []
        elif (self.readOptions=="cached"):
            spec = self.meta_data.getSpectrum(spec_indice)
            with code_block_timer(f'Extracting mz, int, im data arrays...', LOGGER.debug):
                das = self.exp.getSpectrumById(spec_indice).getDataArrays()
                mz_array = das[0].getData()
                int_array = das[1].getData()
                if self.has_im:
                    im_array = das[2].getData()
                else:
                    im_array = []
        else:
            LOGGER.error(f"ERROR: Unknown readOptions ({self.readOptions}) given! Has to be one of 'ondisk', 'cached'")
        
        return (mz_array, int_array, im_array)
 

    def filter_single_spectrum(self, 
                               spec_indice: int, 
                               feature: TransitionGroupFeature,
                               config: TargetedDIAConfig) -> po.MSSpectrum:
        """
        Filter a single spectrum for a given spectrum indice.

        Args:
          spec_indice: (int) an interger of the spectrum indice to extra a spectrum for
          feature: (TransitionGroupFeature) metadata on feature
          config: (TargetedDIAConfig object) an object of TargetedDIAConfig that contains configuration parameters of how wide filtering windows should be

        Return:
          po.MSSpectrum() a MSSpectrum object 
        """
        # create extraction parameters 
        if self.has_im:
            if feature.consensusApexIM is not None:
                im_start, im_end = config.get_im_upper_lower(feature.consensusApexIM)
            else:
                LOGGER.critical(f"ion mobility information not found in feature {feature.sequence} but present in mzML file")
        # Get tolerance bounds on coordinates
        target_precursor_mz_lower, target_precursor_mz_upper = config.get_upper_lower_tol(feature.precursor_mz)
        target_product_upper_lower_list = [config.get_upper_lower_tol(mz) for mz in feature.product_mz]

        spec_meta = self.meta_data.getSpectrum(spec_indice)


        writeSpectrum = False # boolean flag of whether to write out filtered spectrum

        if spec_meta.getMSLevel() == 1 and 1 in config.mslevel:
            mz_array, int_array, im_array = self.load_spectrum(spec_indice)
            # first filter IM
            if self.has_im:
                if (self.readOptions=="ondisk"): 
                    im_match_bool = (im_array[0].get_data() > im_start) & (
                        im_array[0].get_data() < im_end)
                elif (self.readOptions=="cached"):
                    im_match_bool = (im_array > im_start) & (
                        im_array < im_end)
            else:
                im_match_bool = np.ones(mz_array.shape)

            mz_match_bool = (mz_array > target_precursor_mz_lower) & (
                mz_array < target_precursor_mz_upper)

            writeSpectrum = any(mz_match_bool*im_match_bool)
            if writeSpectrum:
                LOGGER.debug(
                    f"Adding MS1 spectrum {spec_meta.getNativeID()} with spectrum indice {spec_indice} filtered for {sum(mz_match_bool*im_match_bool)} spectra between {target_precursor_mz_lower} m/z and {target_precursor_mz_upper} m/z") # and IM between {im_start} and {im_end}")


        elif spec_meta.getMSLevel() == 2 and 2 in config.mslevel:
            # Get SWATH windows upper and lower
            current_prec = spec_meta.getPrecursors()[0]

            swath_mz_win_lower = current_prec.getMZ() - current_prec.getIsolationWindowLowerOffset()
            swath_mz_win_upper = current_prec.getMZ() + current_prec.getIsolationWindowUpperOffset()

            # Only load and extract product spectra if current spectrums isolation window contains precursor mz
            if feature.precursor_mz > swath_mz_win_lower and feature.precursor_mz < swath_mz_win_upper:
                mz_array, int_array, im_array = self.load_spectrum(spec_indice)
                if self.has_im:
                    if (self.readOptions=="ondisk"): 
                        im_match_bool = (im_array[0].get_data() > im_start) & (
                            im_array[0].get_data() < im_end)
                    elif (self.readOptions=="cached"):
                        im_match_bool = (im_array > im_start) & (
                            im_array < im_end)
                else:
                    im_match_bool = np.ones(mz_array.shape)

                mz_match_bool = np.array(list(map(config.is_mz_in_product_mz_tol_window, mz_array, itertools.repeat(
                    target_product_upper_lower_list, len(mz_array)))))
                writeSpectrum = any(mz_match_bool*im_match_bool)
                if writeSpectrum:
                    LOGGER.debug(
                        f"Feature {feature.sequence}{feature.precursor_charge} - Adding MS2 spectrum {spec_meta.getNativeID()} with spectrum indice {spec_indice} filtered for {sum(mz_match_bool*im_match_bool)} spectra between {target_product_upper_lower_list} m/z") #and IM between {im_start} and {im_end}")
            else:
                LOGGER.debug(f"Feature {feature.sequence}{feature.precursor_charge} Skipping MS2 spectrum {spec_meta.getNativeID()} with spectrum indice {spec_indice} because current swath isolation window ({swath_mz_win_lower} m/z - {swath_mz_win_upper} m/z) does not contain target precursor m/z ({feature.consensusApex})")

        # Only write out filtered spectra if there is any fitlered spectra to write out
        spec_out = po.MSSpectrum()
        if writeSpectrum:
            with code_block_timer(f'Getting filtered spectrum...', LOGGER.debug):
                extract_target_indices = np.where(mz_match_bool * im_match_bool)
                filtered_mz = mz_array[extract_target_indices]
                filtered_int = int_array[extract_target_indices]
                if self.has_im:
                    if (self.readOptions=="ondisk"):
                        filtered_im = im_array[0].get_data()[extract_target_indices]
                    elif (self.readOptions=="cached"):
                        filtered_im = im_array[extract_target_indices]
                else:
                    filtered_im = []
                LOGGER.warn(f"INFO: Adding filtered mz data of length {len(filtered_mz)} | int of length {len(filtered_int)} | im of length {len(filtered_im)}")
                # replace peak data with filtered peak data
                spec_out.set_peaks((filtered_mz, filtered_int))
                
                if self.has_im:
                    # repalce float data arrays with filtered ion mobility data
                    fda = po.FloatDataArray()
                    filtered_im_np = np.array(filtered_im).astype(np.float32)
                    fda.set_data(filtered_im_np)
                    fda.setName("Ion Mobility")
                    spec_out.setFloatDataArrays([fda])
            
            spec_out.setMSLevel(spec_meta.getMSLevel())
            spec_out.setRT(spec_meta.getRT())
            # If you have a lot of filtered spectra to return, it becomes memory heavy.
            return spec_out

    @method_timer
    def reduce_spectra(self, feature: TransitionGroupFeature , config: TargetedDIAConfig) -> FeatureMap:
        """
        Main method for filtering raw mzML DIA/diaPASEF data given specific set of coordinates to filter for.

        Args:
            feature: (TransitionGroupFeature) a TransitionGroupFeature object that contains coordinates to filter for
            config: (TargetedDIAConfig object) an object of TargetedDIAConfig that contains configuration parameters of how wide filtering windows should be

        Return:
          FeatureMap: a FeatureMap object that contains filtered spectra
        """

        # Get indices for requested ms level spectra
        LOGGER.debug(f"Extracting indices for MS Levels: {config.mslevel}")

        # Initialize empty MSExperiment container to store filtered data
        filtered = po.MSExperiment()

       
        rt_start, rt_end = config.get_rt_upper_lower(feature.consensusApex)
         
        LOGGER.debug(
            f"Extracting spectra for {feature.sequence}{feature.precursor_charge} | RT: {feature.consensusApex} sec ({rt_start} - {rt_end})")

        # Restrict spectra list further for RT window, to reduce the number of spectra we need to check to perform filtering on
        use_rt_spec_indices = np.where(np.array(
            [spec_rt >= rt_start and spec_rt <= rt_end for spec_rt in self.get_spectra_rt_list()]))
        target_spectra_indices = np.intersect1d(
            self.get_target_ms_level_indices(config.mslevel), use_rt_spec_indices)
        
        with code_block_timer(f"Filtering {target_spectra_indices.shape[0]} Spectra for {feature.sequence}{feature.precursor_charge}...", LOGGER.debug):
            try:
                filt_spec_list = Parallel(n_jobs=1)(delayed(self.filter_single_spectrum)(spectrum_indice, feature, config) for spectrum_indice in target_spectra_indices.tolist())
            except:
                traceback.print_exc(file=sys.stdout)

        # Add filtered spectra to MSExperiment container
        filt_spec_list = list(
            filter(lambda spectra: spectra is not None, filt_spec_list))
        # Add filtered spectrum to filtered MSExperiment container
        _ = [filtered.addSpectrum(spec) for spec in filt_spec_list]

        #return filtered
        return self.msExperimentToFeatureMap(filtered, feature, config)

    # @method_timer
    def msExperimentToFeatureMap(self, msExperiment: po.MSExperiment, feature: TransitionGroupFeature, config: TargetedDIAConfig ) -> FeatureMap:
        """
        Convert filtered spectra to Pandas DataFrame

        Args:
            msExperiment: (MSExperiment) MSExperiment object that contains filtered data
            feature: (TransitionGroupFeature) metadata on feature
            config: (TargetedDIAConfig object) an object of TargetedDIAConfig that contains configuration parameters of how wide filtering windows should be
          
        Return:
            FeatureMap: a FeatureMap object that contains filtered spectra
        """
        results_df = pd.DataFrame()
        for k in tqdm(range(msExperiment.getNrSpectra())):
            spec = msExperiment.getSpectrum(k)
            mz, intensity = spec.get_peaks()
            if (len(mz) == 0 and len(intensity) == 0):
                LOGGER.warn(
                    f"MS{spec.getMSLevel()} spectrum native id {spec.getNativeID()} had no m/z or intensity array, skipping this spectrum")
                continue
            rt = np.full([mz.shape[0]], spec.getRT(), float)
            
            if not self.has_im:
                LOGGER.warn(
                    f"MS{spec.getMSLevel()} spectrum native id {spec.getNativeID()} had no ion mobility array")
                im = np.full([mz.shape[0]], np.nan, float)
            else:
                im_tmp = spec.getFloatDataArrays()[0]
                im = im_tmp.get_data()
                
            LOGGER.debug(
                f"Adding MS{spec.getMSLevel()} spectrum for peptide: {feature.sequence}{feature.precursor_charge} with native id: {spec.getNativeID()}")
            add_df = pd.DataFrame({'native_id': spec.getNativeID(), 'ms_level': spec.getMSLevel(), 'precursor_mz': feature.precursor_mz,
                                   'mz': mz, 'rt': rt, 'im': im, 'int': intensity})
            results_df = pd.concat([results_df, add_df])

        if not results_df.empty:
            ## Add annotation and column
            results_df['Annotation'] = results_df.apply(self._apply_mz_mapping, args=(feature.product_mz, feature.product_annotations) , axis=1)
            
            ## Add product/precursor mz column
            annotation_mz_mapping = pd.DataFrame({'Annotation': feature.product_annotations, 'product_mz': feature.product_mz})
            annotation_mz_mapping = pd.concat([annotation_mz_mapping, pd.DataFrame({'Annotation': ['prec'], 'product_mz': [feature.precursor_mz]})])
            results_df = results_df.merge(annotation_mz_mapping, on='Annotation', how='left')
        else:
            LOGGER.warn(f"No spectra found for peptide: {feature.sequence}{feature.precursor_charge}. Try adjusting the extraction parameters")

        return FeatureMap(results_df, config)
    
    def _find_closest_reference_mz(self, given_mz: np.array, reference_mz_values: np.array, peptide_product_annotation_list: np.array) -> np.array:
        """
        Find the closest reference m/z value in the given list to provided m/z values.
        Args:
            given_mz (np.array): An array of m/z values for which to find the closest reference m/z values.
            reference_mz_values (np.array): An array of reference m/z values to compare against.
            peptide_product_annotation_list (np.array): An array of reference m/z value annotations.
        Returns:
            np.array: An array of the closest reference m/z values annotations from the provided list.
        """
        return peptide_product_annotation_list[np.argmin(np.abs(reference_mz_values - given_mz[:, None]), axis=1)]

    def _apply_mz_mapping(self, row: pd.DataFrame, peptide_product_mz_list: List[float], peptide_product_annotation_list: List[str]) -> Literal["float", "np.nan"]:
        """
        Apply mz mapping to the given row based on the ms_level.
        Args:
            row (pd.DataFrame): The row containing the data.
            peptide_product_mz_list (List): The list of peptide product m/z values.
            peptide_product_annotation_list (List): The list of peptide product annotations.
        Returns:
            Union[float, np.nan]: The mapped m/z value.
        """
        if row['ms_level'] == 2:
            return self._find_closest_reference_mz(np.array([row['mz']]), np.array(peptide_product_mz_list), np.array(peptide_product_annotation_list))[0]
        elif row['ms_level'] == 1:
            return 'prec'
        else:
            raise ValueError(f"Unknown ms_level {row['ms_level']} encountered.")

