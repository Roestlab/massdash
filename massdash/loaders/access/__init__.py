"""
massdash/loaders/access
~~~~~~~~~~~~~~~~~~~~~~~

The :mod:`massdash.loaders.access` subpackage contains the low level access to DIA data files. 
"""

from .GenericResultsAccess import GenericResultsAccess
from .MzMLDataAccess import MzMLDataAccess
from .OSWDataAccess import OSWDataAccess
from .ResultsTSVDataAccess import ResultsTSVDataAccess
from .SqMassDataAccess import SqMassDataAccess
from .TransitionPQPDataAccess import TransitionPQPDataAccess
from .TransitionTSVDataAccess import TransitionTSVDataAccess

__all__ = [ "GenericResultsAccess",
            "MzMLDataAccess",
            "OSWDataAccess",
            "ResultsTSVDataAccess",
            "SqMassDataAccess",
            "TransitionPQPDataAccess",
            "TransitionTSVDataAccess"]