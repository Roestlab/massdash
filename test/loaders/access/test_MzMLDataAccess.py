"""
test/loaders/access/test_MzMLDataAccess
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

import pandas as pd
import numpy as np
from pathlib import Path
import os
import pyopenms as po
import sys

import pytest
from massdash.loaders.access.MzMLDataAccess import MzMLDataAccess
from massdash.structs import TargetedDIAConfig, TransitionGroupFeature
from massdash.testing.PandasSnapshotExtension import PandasSnapshotExtenstion
from massdash.testing.NumpySnapshotExtension import NumpySnapshotExtenstion
from massdash.util import find_git_directory

## Note: Cached is not tested

TEST_PATH = find_git_directory(Path(__file__).resolve()).parent / 'test'

@pytest.fixture
def mzml_data_access():
    # resave the experiment for this specific os
    if sys.platform == 'win32':

        exp = po.MSExperiment()
        po.MzMLFile().load(os.path.join(TEST_PATH, 'test_data', 'mzml', 'ionMobilityTest.mzML'), exp)
        po.MzMLFile().store(os.path.join(TEST_PATH, 'test_data', 'mzml', 'ionMobilityTest_tmp.mzML'), exp)

        mzml_data_access = MzMLDataAccess(os.path.join(TEST_PATH, 'test_data', 'mzml', 'ionMobilityTest_tmp.mzML'), readOptions='ondisk')
        return mzml_data_access
    else:
        mzml_data_access = MzMLDataAccess(os.path.join(TEST_PATH, 'test_data', 'mzml', 'ionMobilityTest.mzML'), readOptions='ondisk')
        return mzml_data_access

@pytest.fixture
def snapshot_pandas(snapshot):
    return snapshot.use_extension(PandasSnapshotExtenstion)

@pytest.fixture
def snapshot_numpy(snapshot):
    return snapshot.use_extension(NumpySnapshotExtenstion)

@pytest.fixture
def peptide_product_annotation_list():
    return np.array(["A", "B", "C", "D", "E"])

@pytest.fixture
def reference_mz_values():
    return np.array([100.0, 150.0, 200.0, 250.0, 300.0])

def test_load_data(mzml_data_access):
    mzml_data_access.load_data()
    assert mzml_data_access.exp is not None
    assert mzml_data_access.meta_data is not None

@pytest.mark.parametrize("mslevel", [[1, 2], [1], [2]])
def test_get_target_ms_level_indices(mzml_data_access, mslevel, snapshot):
    indices = mzml_data_access.get_target_ms_level_indices(mslevel)
    assert snapshot == indices 

def test_get_spectra_rt_list(mzml_data_access, snapshot):
    rt_list = mzml_data_access.get_spectra_rt_list()
    assert snapshot == rt_list

@pytest.mark.parametrize("spec_indice", [0])
def test_load_spectrum(mzml_data_access, spec_indice, snapshot_numpy):
    mz, intens, im = mzml_data_access.load_spectrum(spec_indice)
    assert snapshot_numpy == np.column_stack([mz, intens, im[0].get_data()]) # note IM not tested here

@pytest.mark.parametrize("spec_indice", [0])
def test_filter_single_spectrum(mzml_data_access, spec_indice, snapshot_numpy):
    product_mzs = [504.2664, 591.2984, 704.3825, 851.4509, 966.4779, 1065.5463]
    annotations = ['y4^1', 'y5^1', 'y6^1', 'y7^1', 'y8^1', 'y9^1'] 
    feature = TransitionGroupFeature(leftBoundary=6233, 
                                     rightBoundary=6247, 
                                     consensusApex=6240.41, 
                                     consensusApexIM=0.98, 
                                     sequence='AFVDFLSDEIK', 
                                     precursor_charge=2, 
                                     precursor_mz=642.3295,
                                     product_mz=product_mzs,
                                     product_annotations=annotations)

    config = TargetedDIAConfig()
    filtered_spectrum = mzml_data_access.filter_single_spectrum(spec_indice, feature, config)
    fda = filtered_spectrum.getFloatDataArrays()[0]
    im = fda.get_data()
    mz, intens = filtered_spectrum.get_peaks()
    peaks = np.column_stack([mz, intens, im])
    assert snapshot_numpy == peaks 

def test_reduce_spectra(mzml_data_access, snapshot_pandas): # also tests msExperimentToFeatureMap()
    product_mzs = [504.2664, 591.2984, 704.3825, 851.4509, 966.4779, 1065.5463]
    annotations = ['y4^1', 'y5^1', 'y6^1', 'y7^1', 'y8^1', 'y9^1'] 
    feature = TransitionGroupFeature(leftBoundary=6233, 
                                     rightBoundary=6247, 
                                     consensusApex=6240.41, 
                                     consensusApexIM=0.98, 
                                     sequence='AFVDFLSDEIK', 
                                     precursor_charge=2, 
                                     precursor_mz=642.3295,
                                     product_mz=product_mzs,
                                     product_annotations=annotations)
    config = TargetedDIAConfig()
    config.rt_window = 2
    feature_map = mzml_data_access.reduce_spectra(feature, config)
    assert snapshot_pandas == feature_map.feature_df

@pytest.mark.parametrize("mz,expected_annot", [(150.01, 'B'), (249.99, 'D')])
def test_find_closest_reference_mz(reference_mz_values, peptide_product_annotation_list, mz, expected_annot):
    closest_mz_annot = MzMLDataAccess._find_closest_reference_mz(mz, reference_mz_values, peptide_product_annotation_list)
    assert expected_annot == closest_mz_annot

@pytest.mark.parametrize("level,mz,expected_annot", [(1, 100, 'prec'), (2, 200.01, 'C'), (2, 299.89, 'E')])
def test_apply_mz_mapping(reference_mz_values, peptide_product_annotation_list, level, mz, expected_annot):
    row = pd.Series(dict(mz=mz, ms_level=level))
    result = MzMLDataAccess._apply_mz_mapping(row, reference_mz_values, peptide_product_annotation_list)
    assert expected_annot == result