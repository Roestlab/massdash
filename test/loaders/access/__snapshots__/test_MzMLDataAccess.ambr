# serializer version: 1
# name: test_get_spectra_rt_list
  array([6225.00510572, 6225.11081679, 6226.79294987, 6226.89773809,
         6228.58093243, 6228.68570364, 6230.36718894, 6230.47519996,
         6232.15643628, 6232.26158461, 6233.94503701, 6234.05347881,
         6235.73636055, 6235.84861018, 6237.52955912, 6237.64613564,
         6239.32257002, 6239.43720522, 6241.11678499, 6241.24574655,
         6242.92556385, 6243.03476927, 6244.714715  , 6244.82797253,
         6246.5069831 , 6246.62867474, 6248.30696891, 6248.42832918,
         6250.10570723, 6250.22954839, 6251.90983957, 6252.03127199,
         6253.71197694, 6253.83602161, 6255.51902593, 6255.64580925,
         6257.32368641, 6257.45004012, 6259.12679414, 6259.29275457,
         6260.96995081, 6261.10140645, 6262.77919124, 6262.90909502,
         6264.5863851 , 6264.71157343, 6266.3903799 , 6266.5151357 ,
         6268.1971333 , 6268.33268181])
# ---
# name: test_get_target_ms_level_indices[mslevel0]
  array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
         17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
         34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])
# ---
# name: test_get_target_ms_level_indices[mslevel1]
  array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32,
         34, 36, 38, 40, 42, 44, 46, 48])
# ---
# name: test_get_target_ms_level_indices[mslevel2]
  array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33,
         35, 37, 39, 41, 43, 45, 47, 49])
# ---
